lib/Stomp/Server.pm6

```diff
@@ -1,26 +1,106 @@
+
+    class Subscription {
+        has Str $.id            is required;
+        has Str $.destination   is required;
+        has AckMode $.ack = 'auto';
+
+        method ACCEPTS(Stomp::Message $mess) {
+            $!destination eq $mess.headers<destination>;
+        }
+    }
+
+    class Connection does Stomp::MessageStream[Stomp::Parser::ClientCommands] {
+        has $.conn;
+        has Supply $!messages;
+
+        has  Subscription @.subscriptions;
+        has  Lock $!subscription-lock;
+
+        has Supplier $!sent-message-supplier;
+
+        submethod TWEAK {
+            $!messages = self!process-messages($!conn.Supply).share;
+
+            $!subscription-lock = Lock.new;
+
+            my &quit = {
+                when X::Stomp::MalformedMessage {
+                    await $!conn.print: Stomp::Message.new:
+                                command => 'ERROR',
+                                body => .message;
+                }
+            };
+
+            my $connect-tap = $!messages.grep({ $_.command ~~ 'CONNECT'|'STOMP' }).tap: 
                     {
-                        await $conn.print: Stomp::Message.new:
+                        await $!conn.print: Stomp::Message.new:
                             command => 'CONNECTED',
                             headers => ( accept-version => '1.2' );
-                    },
-                    quit => {
-                        when X::Stomp::MalformedMessage {
-                            await $conn.print: Stomp::Message.new:
-                                command => 'ERROR',
-                                body => .message;
-                        }
-                    };
+                        $connect-tap.close;
+                    }, :&quit;
+            $!messages.grep({ $_.command ~~ 'SUBSCRIBE' }).tap: {
+                $!subscription-lock.protect: {
+                    @!subscriptions.push: Subscription.new( id          => $_.headers<id>, 
+                                                            destination => $_.headers<destination>,
+                                                            ack         => $_.headers<ack> // 'auto' );
+                };
+            }, :&quit;
+            $!messages.grep({ $_.command ~~ 'UNSUBSCRIBE' }).tap: {
+                $!subscription-lock.protect: {
+                    my $id = $_.headers<id>;
+                    @!subscriptions = @!subscriptions.grep({ $_.id !~~ $id });
+                };
+            }, :&quit;
+
+            $!sent-message-supplier = Supplier.new;
+            $!messages.grep({ $_.command ~~ 'SEND' }).tap: {
+                $!sent-message-supplier.emit: $_;
+            }, :&quit;
+        }
+
+        method sent-messages() returns Supply {
+            $!sent-message-supplier.Supply;
+        }
+
+        method ACCEPTS(Stomp::Message $mess) {
+            @!subscriptions.map(*.destination).any eq $mess.headers<destination>;
```

Comment:
This access to `@!subscriptions` is not protected by the lock, so it may be read concurrent with modification.